File: app/src/main/java/ru/voronezhtsev/weatherapp/di/WeatherComponent.java
package ru.voronezhtsev.weatherapp.di;

import android.content.Context;

import javax.inject.Singleton;

import dagger.BindsInstance;
import dagger.Component;
import ru.voronezhtsev.weatherapp.data.repositories.ForecastsRepository;
import ru.voronezhtsev.weatherapp.domain.ICityRepository;
import ru.voronezhtsev.weatherapp.domain.WeatherInteractor;
import ru.voronezhtsev.weatherapp.presentation.CityChoicePresenter;
import ru.voronezhtsev.weatherapp.presentation.MainPresenter;

@Singleton
@Component(modules = {WeatherModule.class})
public interface WeatherComponent {


    ForecastsRepository getForecastsRepository();

    WeatherInteractor getWeatherInteractor();

    ICityRepository getCityRepository();

    MainPresenterFactory getMainPresenterFactory();

    CityChoicePresenter getCityChoicePresenter();

    @Component.Builder
    interface Builder {
        WeatherComponent build();
        @BindsInstance
        Builder component(Context context);
    }
}

File: app/src/main/java/ru/voronezhtsev/weatherapp/di/WeatherComponent.kt
package ru.voronezhtsev.weatherapp.di

import android.content.Context
import dagger.BindsInstance
import dagger.Component
import ru.voronezhtsev.weatherapp.presentation.MainScreenViewModelFactory
import ru.voronezhtsev.weatherapp.presentation.forecast.ForecastViewModelFactory
import ru.voronezhtsev.weatherapp.presentation.weatherlist.WeatherListViewModelFactory
import javax.inject.Singleton

@Singleton
@Component(modules = [WeatherModule::class])
interface WeatherComponent {
    val weatherListViewModelFactory: WeatherListViewModelFactory
    val mainScreenViewModelFactory: MainScreenViewModelFactory
    val forecastViewModelFactory: ForecastViewModelFactory

    @Component.Builder
    interface Builder {
        fun build(): WeatherComponent
        @BindsInstance
        fun component(context: Context): Builder
    }
}